src/renderer/contexts/DialogContext.tsx

```diff
@@ -0,0 +1,91 @@
+import React, { createElement, useCallback, useMemo, useState } from 'react'
+
+import type { JSXElementConstructor, PropsWithChildren } from 'react'
+
+import { generateRandomUUID } from '../utils/random'
+
+type DialogId = string
+
+export type DialogProps = {
+  onClose: () => void
+}
+
+type DialogElementConstructor<T> = JSXElementConstructor<DialogProps & T>
+
+export type OpenDialog = <T extends { [key: string]: any }>(
+  dialogElement: DialogElementConstructor<T>,
+  additionalProps?: T
+) => DialogId
+
+export type CloseDialog = (id: DialogId) => void
+
+type DialogContextValue = {
+  hasOpenDialogs: boolean
+  openDialog: OpenDialog
+  closeDialog: CloseDialog
+}
+
+const initialValues: DialogContextValue = {
+  hasOpenDialogs: false,
+  openDialog: _ => '',
+  closeDialog: _ => {},
+}
+
+export const DialogContext = React.createContext<DialogContextValue>(
+  initialValues
+)
+
+export const DialogContextProvider = ({ children }: PropsWithChildren<{}>) => {
+  const [dialogs, setDialogs] = useState<{ [id: DialogId]: JSX.Element }>({})
```

Comment:
I feel this can be written a bit more readable. How about `Map<DialogId, JSX.Element>`?

I know Map and Object are not the same. But why an object is used here rather than a map?

Reply:
`Record<keytype,valuetype>` can be used for objects https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type

I personally also of the use that syntax (`{[key: type]: type}`), so it's fine for me, I don't really have a preference between the two but maybe `Record` is better?

Reply:
> I feel this can be written a bit more readable. How about `Map<DialogId, JSX.Element>`?
> 
> I know Map and Object are not the same. But why an object is used here rather than a map?

Maps are really quite different than objects and we're not using them as far as I know (you can't just use `{}` as an replacement of a Map).

`Record` as @treefit suggests is the closest replacement which would work.

I'm also open to any way of writing this type down (Record or `{}` syntax), happy to change it in whatever way we decide and stick to it.

Reply:
I strongly suggest `Record` as it's more readable and feels more standard.

Reply:
note that I think it's not worth to discuss about this small cosmetic thing here...